### PR TITLE
Create algorithm to find all possible meeting times

### PR DESCRIPTION
**Pull request by shreyachatterjee00**
Passes all pre-written tests: 
* Functionality for finding a meeting time given a meeting request and list of events on that day 
* Works for conflicting events, nested events, no events, etc

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 1]
```diff
@@ -34,7 +34,7 @@
 
 /** Servlet that deletes bucket list content*/
 @WebServlet("/delete-data")
-public class DeleteDataServlet extends HttpServlet 
+public class DeleteDataServlet extends HttpServlet {
```
Comment (megankj) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:37]: This file + style.css look unrelated to this PR.  Was this an unintentional commit?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,96 @@
...
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.ArrayList;
+import java.util.Collections;
+
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Please add a javadoc comment explaining what this method does.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,96 @@
...
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long mtngDuration = request.getDuration();
+    Collection<String> attendees = request.getAttendees();
+    Collection<Event> allEvents = events;
+
+    /* Return an array list of Time Range objects that span any available time. For example, if 4pm - 8pm is free for a meeting, 
+      the Time Range object will represent that entire chunk of time rather than 4pm - 5pm, even if the duration of meeting is 1 hour. */
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Lines for java files should not be longer than 100 characters.  Multi-line comments should be in this form:

```
  /**
   * multi line comment that is shorter than 100 characters
   * that goes on to the next line
   */
```

Please update here and throughout.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,96 @@
...
+
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long mtngDuration = request.getDuration();
+    Collection<String> attendees = request.getAttendees();
+    Collection<Event> allEvents = events;
+
+    /* Return an array list of Time Range objects that span any available time. For example, if 4pm - 8pm is free for a meeting,
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Is this the comment for the entire method?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,96 @@
...
+              longestMeetingEnd = meetingEnd;
+            }
+          }
+          tempTime += 10;
+        }
+          
+      currTime = longestMeetingEnd;
+      startTime = longestMeetingEnd;
+      }
+
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:105]: Don't orphan else statements like this, it might be hard to read!
```suggestion
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -15,9 +15,96 @@
...
+      /* If there is no event at this time, add 30 minutes to the duration and current time, and run the loop again. */
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: 30 or 10?  Also it's better to make these constants so they're easier to update and easier to read.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -15,9 +15,96 @@
...
+            int meetingEnd = Collections.max(conflictMeetingEndTimes);
+            if (meetingEnd > longestMeetingEnd) {
+              longestMeetingEnd = meetingEnd;
+            }
+          }
+          tempTime += 10;
+        }
+          
+      currTime = longestMeetingEnd;
+      startTime = longestMeetingEnd;
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
        currTime = longestMeetingEnd;
        startTime = longestMeetingEnd;
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -15,9 +15,96 @@
...
+      /* If there is no event at this time, add 30 minutes to the duration and current time, and run the loop again. */
+      else {
+        currDuration += 10;
+        currTime +=10;
+       }
+    }
+
+    /* When reaching the end of loop, if the current duration is longer than what is necessary, add it to the return Collection, inclusive. */
+    if (currTime > TimeRange.END_OF_DAY) {
+      if (currDuration >= mtngDuration) {
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
    if (currTime > TimeRange.END_OF_DAY && currDuration >= mtngDuration) {
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -15,9 +15,96 @@
...
+        meetingTimes.add(TimeRange.fromStartEnd(startTime, TimeRange.END_OF_DAY, true));
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: does it end at END_OF_DAY because TimeRange can't go into the next day?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -15,9 +15,96 @@
...
+      currTime = longestMeetingEnd;
+      startTime = longestMeetingEnd;
+      }
+
+      /* If there is no event at this time, add 30 minutes to the duration and current time, and run the loop again. */
+      else {
+        currDuration += 10;
+        currTime +=10;
+       }
+    }
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:109]: This is looking good, but it's pretty long for one method so I wonder if you can break this up into more logical helper methods.  From the looks of it, some of the steps you're doing are:
1) setting up the map of event times w/ conflicts
2) processing the map to determine where the "free time" is in the map 

Maybe at least those can be broken up?  Smaller helper methods will also help document the code.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -15,9 +15,96 @@
...
+    int startTime = TimeRange.START_OF_DAY;
+    int currDuration = 0;
+
+    while (currTime < TimeRange.END_OF_DAY) {
+      /* If there's an event starting at this time, compare the current duration and meeting duration to see if there is a possible meeting time. 
+        Reset the current duration variable, move the current time to the end time of the event, and repeat loop. */
+      if (mandatoryEvents.containsKey(currTime)) {
+        ArrayList<Integer> endTimes = mandatoryEvents.get(currTime);
+        // Get list item with latest end time
+        int longestMeetingEnd = Collections.max(endTimes);
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I may be wrong, but it looks like you only ever care about the longest meeting for a given start time.  Do you need to store the other ones?  Could that be pre-processed?  (Or does that increase complexity?)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -15,9 +15,96 @@
...
+    while (currTime < TimeRange.END_OF_DAY) {
+      /* If there's an event starting at this time, compare the current duration and meeting duration to see if there is a possible meeting time. 
+        Reset the current duration variable, move the current time to the end time of the event, and repeat loop. */
+      if (mandatoryEvents.containsKey(currTime)) {
+        ArrayList<Integer> endTimes = mandatoryEvents.get(currTime);
+        // Get list item with latest end time
+        int longestMeetingEnd = Collections.max(endTimes); 
+        if (currDuration >= mtngDuration) {
+            // Not inclusive because a meeting starts at this time 
+            meetingTimes.add(TimeRange.fromStartEnd(startTime, currTime, false));
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: To make it clear what this boolean is, I suggest adding a comment like this:
```suggestion
            meetingTimes.add(TimeRange.fromStartEnd(startTime, currTime, /* inclusive= */ false));
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 13]
```diff
@@ -15,9 +15,96 @@
...
+        }
+          
+      currTime = longestMeetingEnd;
+      startTime = longestMeetingEnd;
+      }
+
+      /* If there is no event at this time, add 30 minutes to the duration and current time, and run the loop again. */
+      else {
+        currDuration += 10;
+        currTime +=10;
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
        currTime += 10;
```

**Review by megankj — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java (hunk 1 above)
Comment (shreyachatterjee00) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:37]: So I had just merged multiple other PRs into master, and then pulled those before starting development on this branch, so I thought I'd make the few design changes that emerged from merging those PRs, on this branch. I thought it wouldn't be worth it to make a whole new branch for a couple lines of code, but I probably just should have because yeah, this has nothing to do with the rest of the PR.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (shreyachatterjee00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yeah, they provided START_OF_DAY and END_OF_DAY constants which are what the meetings are supposed to fall within.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java (hunk 1 above)
Comment (megankj) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:37]: alright, that's fine. thanks for the explanation!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 14]
```diff
@@ -15,9 +15,120 @@
...
 
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.ArrayList;
+import java.util.Collections;
 
+/** 
+* Returns a Collection of TimeRange objects that span any available meeting times for all people in the MeetingRequest request.
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: use javadoc formatting {@code TimeRange} which will help link all these objects together when looking at the codebase in github!

Comment (megankj) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: also this javadoc should probably go above the query method, not the entire class

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 15]
```diff
@@ -15,9 +15,120 @@
...
 public final class FindMeetingQuery {
+  public static final int MIN_INCREMENT = 10;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long mtngDuration = request.getDuration();
+    Collection<String> attendees = request.getAttendees();
+    Collection<Event> allEvents = events;
+
+    HashMap<Integer, Integer> mandatoryEvents = makeMandatoryEventMap(allEvents, attendees);
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I think it would be good to explain what mandatory means

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 16]
```diff
@@ -15,9 +15,120 @@
...
+  * Create a hash map <event start time, event end time> that ONLY contains meetings where 1 or more people from attendees must attend. 
+  **/
+  public HashMap<Integer, Integer> makeMandatoryEventMap (Collection<Event> allEvents, Collection<String> attendees) {
+    HashMap<Integer, Integer> mandatoryEvents = new HashMap<Integer, Integer>();
+
+    for (Event event : allEvents) {
+      //people who must attend 
+      Set<String> people = event.getAttendees();
+      //check if any people in meeting request is present in this set
+      for (String person : attendees) {
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I don't think you actually need to iterate through every single attendee, right?  You just need to determine if this event contains any of the provided attendees, and if so, add the event time to your map.

You may be able to remove some of this nesting by using [java streams anyMatch](https://www.geeksforgeeks.org/stream-anymatch-java-examples/) to find out of this event is potentially conflicting with the request.  anyMatch is nice because it's short-circuiting and will terminate if it finds a match, not having to go through all attendees.

boolean eventHasAttendee = attendees.stream().anyMatch(attendee -> event.getAttendees().contains(attendee));

if (eventHasAttendee) {
  addMandatoryEvent(event);
}

And then your addMandatory event function can add the event to your map (unless it's shorter than the existing entry).

(As a side note, code with lots of nesting has the potential to introduce bugs since it makes it hard to follow the conditional branching that may or may not occur)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 17]
```diff
@@ -15,9 +15,120 @@
...
+          }
+        }  
+      }
+    }
+
+    return mandatoryEvents;
+  }
+
+  /**
+  * Create and return an ArrayList of TimeRange objects that span available meeting times.
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: Method comments should be in the form "Creates and returns.." instead of "Create and return"

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 18]
```diff
@@ -15,9 +15,120 @@
...
+  */
+  public Collection<TimeRange> findMeetingTimes (long mtngDuration, HashMap<Integer, Integer> mandatoryEvents) {
+    Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+    int currTime = TimeRange.START_OF_DAY;
+    int startTime = TimeRange.START_OF_DAY;
+    int currDuration = 0;
+
+    while (currTime < TimeRange.END_OF_DAY) {
+      // If there's an event starting at this time, compare the current duration and meeting duration to see if there is a possible meeting time.
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: make this comment shorter than 100 characters so doesn't wrap the line like this

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 19]
```diff
@@ -15,9 +15,120 @@
...
+    Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+    int currTime = TimeRange.START_OF_DAY;
+    int startTime = TimeRange.START_OF_DAY;
+    int currDuration = 0;
+
+    while (currTime < TimeRange.END_OF_DAY) {
+      // If there's an event starting at this time, compare the current duration and meeting duration to see if there is a possible meeting time. 
+      // Reset the current duration variable, move the current time to the end time of the event, and repeat loop. 
+      if (mandatoryEvents.containsKey(currTime)) {
+        int longestMeetingEnd = mandatoryEvents.get(currTime);
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Does need to be declared here?  It's immediately re-assigned on line 97.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 20]
```diff
@@ -15,9 +15,120 @@
...
+        if (currDuration >= mtngDuration) {
+          // Not inclusive because a meeting starts at this time 
+          meetingTimes.add(TimeRange.fromStartEnd(startTime, currTime, /* inclusive = */ false));
+        }
+        currDuration = 0;
+
+        // Check if there is a conflict: another meeting that starts during this one, and goes later
+        longestMeetingEnd = isConflict(currTime, longestMeetingEnd, mandatoryEvents);
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Methods with names like isFoo are expected to return booleans.  Instead, since this returns an int, it should be something like getLastEventEnd or something that better describes what int you're getting here.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 21]
```diff
@@ -15,9 +15,120 @@
...
+            
+        currTime = longestMeetingEnd;
+        startTime = longestMeetingEnd;
+      }
+      // If there is no event at this time, add MIN_INCREMENT minutes to the duration and current time, and run the loop again. 
+      else {
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I mentioned this in our 1:1, but it's better to keep the code a little closer together like this:
```suggestion
      } else {
      // If there is no event at this time, add MIN_INCREMENT minutes to the duration and current time, and run the loop again. 
```

Please update here and elsewhere.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 22]
```diff
@@ -15,9 +15,120 @@
...
+    }
+    return meetingTimes;
+  }
+
+  /**
+  * Check if there is a conflict: another meeting that starts during this one, and goes later. 
+  * Return the later time if there is a conflicting meeting. 
+  **/
+  public int isConflict (int currTime, int longestMeetingEnd, HashMap<Integer, Integer> mandatoryEvents) {
+    int tempTime = currTime + 10;
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: use MIN_INCREMENT?

**Review by megankj — CHANGES_REQUESTED** [review]
looking good, thanks for making the changes!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 19 above)
Comment (shreyachatterjee00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I  need to keep track of the current longest meeting end because I'm comparing against it in the isConflict function, but I can call this variable something else to make it clearer. Instead of making this a variable at all I could also just pass "mandatoryEvents.get(currTime)" into the function directly, but I feel like that's not as clean

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 23]
```diff
@@ -20,61 +20,56 @@
...
-* the Time Range object will represent that entire chunk of time rather than 4pm - 5pm, even if the duration of meeting is 1 hour. 
-*/
 public final class FindMeetingQuery {
   public static final int MIN_INCREMENT = 10;
 
+  /** 
+  * Returns a Collection of {@code TimeRange} objects that span any available meeting times for all people in the MeetingRequest request.
+  * For example, if 4pm - 8pm is free for a meeting, 
+  * the Time Range object will represent that entire chunk of time rather than 4pm - 5pm, even if the duration of meeting is 1 hour. 
+  */
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:35]: can you reformat this comment so it's less than 100 chars per line?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 24]
```diff
@@ -20,61 +20,56 @@
...
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
     long mtngDuration = request.getDuration();
     Collection<String> attendees = request.getAttendees();
     Collection<Event> allEvents = events;
 
+    // Mandatory events = events that must be attended by an attended in the Meeting Request
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
    // Mandatory events are events that must be attended by an attendee in the {@code MeetingRequest}.
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 25]
```diff
@@ -15,9 +15,114 @@
...
+  /** 
+  * Returns a Collection of {@code TimeRange} objects that span any available meeting times for all people in the MeetingRequest request.
+  * For example, if 4pm - 8pm is free for a meeting, 
+  * the Time Range object will represent that entire chunk of time rather than 4pm - 5pm, even if the duration of meeting is 1 hour. 
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long mtngDuration = request.getDuration();
+    Collection<String> attendees = request.getAttendees();
+    Collection<Event> allEvents = events;
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: is there a reason you make a copy of this?

Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: this is a small nit, but none of these need to be local variables, you can just inline them in the makeMandatoryEventMap and findMeetingTimes methods below

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 26]
```diff
@@ -15,9 +15,114 @@
...
+        } else {
+          mandatoryEvents.put(eventStartTime, eventEndTime);
+        }
+      }
+    }
+    return mandatoryEvents;
+  }
+
+  /**
+  * Creates and returns an ArrayList of TimeRange objects that span available meeting times.
```
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
  * Creates and returns an ArrayList of {@code TimeRange} objects that span available meeting times.
```

**Review by megankj — CHANGES_REQUESTED** [review]
Does running "make pretty" format java files?  I'm wondering if it will auto-truncate lines to 100 characters for you.

@tjgoog feel free to also leave comments at this point.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 25 above)
Comment (tjgoog) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: To get technical here... this is a new reference pointing to an existing object, so if you do want to make a copy that you can modify without affecting the original (which is a good practice if you are modifying input data), you would do something like
`LinkedList<Event> modifiableEvents = new LinkedList<>(events);`

If you don't make a copy, this is effectively creating an alias for `events`, so if you prefer a different name, just change the parameter name at the top of the method.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 27]
```diff
@@ -15,9 +15,114 @@
...
+    Collection<Event> allEvents = events;
+
+    // Mandatory events = events that must be attended by an attended in the Meeting Request
+    HashMap<Integer, Integer> mandatoryEvents = makeMandatoryEventMap(allEvents, attendees);
+    Collection<TimeRange> meetingTimes = findMeetingTimes(mtngDuration, mandatoryEvents);
+    return meetingTimes;
+  }
+
+  /** 
+  * Creates a hash map <event start time, event end time> that ONLY contains meetings where 1 or more people from attendees must attend.
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Document that if there are two events at the same start time, the end time stored in the map will be the latest end time for that start time.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 28]
```diff
@@ -15,9 +15,114 @@
...
+
+  /** 
+  * Creates a hash map <event start time, event end time> that ONLY contains meetings where 1 or more people from attendees must attend. 
+  **/
+  public HashMap<Integer, Integer> makeMandatoryEventMap (Collection<Event> allEvents, Collection<String> attendees) {
+    HashMap<Integer, Integer> mandatoryEvents = new HashMap<Integer, Integer>();
+
+    for (Event event : allEvents) {
+      //check if any people in meeting request is present in this set
+      boolean mtngAttendeeInEvent = attendees.stream().anyMatch(attendee -> event.getAttendees().contains(attendee));
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:55]: Yay Streams! I like them more than I probably should. (Just a comment, not requested change)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 29]
```diff
@@ -15,9 +15,114 @@
...
+        }
+      }
+    }
+    return mandatoryEvents;
+  }
+
+  /**
+  * Creates and returns an ArrayList of TimeRange objects that span available meeting times.
+  */
+  public Collection<TimeRange> findMeetingTimes (long mtngDuration, HashMap<Integer, Integer> mandatoryEvents) {
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Since the javadoc is saying that this will return an ArrayList, should change method declaration to be ArrayList<TimeRange> or change doc to be Collection.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 26 above)
Comment (tjgoog) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Even better: Use `{@link TimeRange}` so that relevant tools will make this a link to the TimeRange class definition. Also, `{@link }`-ify `ArrayList`.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 30]
```diff
@@ -15,9 +15,114 @@
...
+    int tempTime = currTime + MIN_INCREMENT;
+    int currentEventEnd = longestMeetingEnd;
+    while (tempTime < currentEventEnd) {
+      if (mandatoryEvents.containsKey(tempTime)) {
+        int meetingEnd = mandatoryEvents.get(tempTime);
+        if (meetingEnd > longestMeetingEnd) {
+          longestMeetingEnd = meetingEnd;
+        }
+      }
+      tempTime += 10;
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
      tempTime += MIN_INCREMENT;
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 31]
```diff
@@ -15,9 +15,114 @@
...
+        startTime = longestMeetingEnd;
+      } else {
+        // If there is no event at this time, add MIN_INCREMENT minutes to the duration and current time, and run the loop again. 
+        currDuration += MIN_INCREMENT;
+        currTime += MIN_INCREMENT;
+      }
+    }
+
+    // When reaching the end of loop, if the current duration is longer than what is necessary, add it to the return Collection, inclusive. 
+    if (currTime > TimeRange.END_OF_DAY && currDuration >= mtngDuration) {
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: When execution gets here, `currTime` could be (will be, assuming `END_OF_DAY` is a multiple of `MIN_INCREMENT`, I think) == `END_OF_DAY`. So shouldn't the first comparison be `>=`?

**Review by tjgoog — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 31 above)
Comment (shreyachatterjee00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: yeah, should be

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 32]
```diff
@@ -27,26 +27,29 @@
...
   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    long mtngDuration = request.getDuration();
-    Collection<String> attendees = request.getAttendees();
-    Collection<Event> allEvents = events;
-
-    // Mandatory events = events that must be attended by an attended in the Meeting Request
-    HashMap<Integer, Integer> mandatoryEvents = makeMandatoryEventMap(allEvents, attendees);
-    Collection<TimeRange> meetingTimes = findMeetingTimes(mtngDuration, mandatoryEvents);
+    // Mandatory events are events that must be attended by an attended in the {@link Meeting Request}
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
    // Mandatory events are events that must be attended by a mandatory attendee in the {@link Meeting Request}
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 33]
```diff
@@ -27,26 +27,29 @@
...
+  * Creates a hash map <event start time, event end time> that ONLY contains meetings 
+  * where 1 or more people from attendees must attend. 
+  * If there are two meetings with the same start time,
+  * only the one with a later end time is saved.
   **/
-  public HashMap<Integer, Integer> makeMandatoryEventMap (Collection<Event> allEvents, Collection<String> attendees) {
+  public HashMap<Integer, Integer> makeMandatoryEventMap (Collection<Event> events, MeetingRequest request) {
     HashMap<Integer, Integer> mandatoryEvents = new HashMap<Integer, Integer>();
+    Collection<String> attendees = request.getAttendees();
+    ArrayList<Event> allEvents = new ArrayList<>(events);
```
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Because you're not modifying anything in `events`, no need to create a copy of it, so I think this should just be removed and use `events` in the `for` loop.

**Review by tjgoog — APPROVED** [review]
Approved! Please address my small comments and get Megan's approval too before merging.

**Review by megankj — APPROVED** [review]
